lib/charms/opensearch/v0/opensearch_base_charm.py

```diff
@@ -367,95 +448,157 @@ def _is_tls_fully_configured(self) -> bool:
+        if self.peers_data.has(Scope.UNIT, "update-node-config"):
+            updated_node_conf = Node.from_dict(
+                self.peers_data.get_object(Scope.UNIT, "update-node-config")
+            )
+
         try:
             # Retrieve the nodes of the cluster, needed to configure this node
-            nodes = self._get_nodes()
+            nodes = self._get_nodes(False)
+
+            # Set the configuration of the node
+            self._set_node_conf(nodes, updated_node_conf.roles if updated_node_conf else None)
         except OpenSearchHttpError:
             event.defer()
+            self.peers_data.delete(Scope.UNIT, "starting")
             return
 
-        # Set the configuration of the node
-        self._set_node_conf(nodes)
-
         try:
-            self.unit.status = BlockedStatus(WaitingToStart)
             self.opensearch.start()
+            self._post_start_init()
             self.status.clear(WaitingToStart)
-        except OpenSearchStartError as e:
-            logger.debug(e)
+        except OpenSearchStartTimeoutError:
+            event.defer()
+        except OpenSearchStartError:
+            self.peers_data.delete(Scope.UNIT, "starting")
             self.unit.status = BlockedStatus(ServiceStartError)
             event.defer()
-            return
 
-        # initialize the security index if the admin certs are written on disk
+    def _post_start_init(self):
+        """Initialization post OpenSearch start."""
         if self.unit.is_leader():
+            # initialize the security index if needed and if the admin certs are written on disk
             if not self.peers_data.get(Scope.APP, "security_index_initialised"):
                 admin_secrets = self.secrets.get_object(Scope.APP, CertType.APP_ADMIN.val)
                 self._initialize_security_index(admin_secrets)
                 self.peers_data.put(Scope.APP, "security_index_initialised", True)
 
             self.peers_data.put(Scope.APP, "leader_ip", self.unit_ip)
 
-        # store the exclusions that previously failed to be stored when no units online
-        if self.peers_data.get(Scope.APP, "remove_from_allocation_exclusions"):
-            self.opensearch.remove_allocation_exclusions(
-                self.peers_data.get(Scope.APP, "remove_from_allocation_exclusions")
-            )
-
         # cleanup bootstrap conf in the node
         if self.peers_data.get(Scope.UNIT, "bootstrap_contributor"):
             self._cleanup_bootstrap_conf_if_applies()
 
-    def _stop_opensearch(self, event: EventBase) -> None:
+        # Remove the exclusions that could not be removed when no units were online
+        self.opensearch_exclusions.delete_current()
+
+        # Remove the 'starting' flag on the unit
+        self.peers_data.delete(Scope.UNIT, "starting")
+
+        # set "update complete flag" to true if it was an update
+        current_update_node_conf = self.peers_data.get_object(Scope.UNIT, "update-node-config")
```

Comment:
Node-config logic is overcomplicated. It is simpler to render the config and the current with the expected.

Reply:
Could you elaborate on what you mean by render the config? 
So the flow here:
- leader unit computes the "new node roles" + "target unit where to apply this config": `{"name": "opensearch-1, "roles": [new roles]}`
- it sends it on the peer relation
- `_on_peer_relation_changed`: all units gets the event, but only the target one will pick it up based on `name`
- the target unit will put this updated conf its `unit data bag` 
- it will then emit a `_restart` event done by rollingops
- the `_start_opensearch` method will fire at some point and pick this new conf and render it in its yaml files and restart
- but `_peer_relation_changed` will fire multiple times, so we need to make sure this only updates once
- so when the target unit applied the conf  and restarted it will put the latest config (roles) in its unit data
- when another `_peer_relation_changed` event fires, it will check if its again the target unit (based on the name) and will compare the update payload with the `latest_update_config` and will skip if they're equal